gpu_plugin: add kustomizations

askervin opened the pull request:
- Default deployment: `kubectl apply -k deployments/gpu_plugin`
- Variant: deploy only on nodes with Intel GPU label by NFD
  `kubectl apply -k deployments/gpu_plugin/overlays/nfd_labeled_nodes`
- GPU plugin README updated.

Signed-off-by: Antti Kervinen <antti.kervinen@intel.com>

grahamwhaley commented on cmd/gpu_plugin/README.md:
````diff
@@ -84,10 +84,19 @@ Successfully tagged intel/intel-gpu-plugin:devel
+```
+
+If your cluster runs [Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery),
+you can deploy the device plugin DaemonSet only on nodes with Intel GPU:
````
Given you added the link to the YAML file in the previous paragraph (thanks!), maybe you can link out here to the add-nodeselector yaml file?

grahamwhaley replied:
I also wonder if we should mention `kustomization` here - I'm not sure how common kustomize use is yet, so we might want to call it out to be clear that is in use to do the modification/specialisation of the yaml.

askervin replied:
Added link to the kustomization.
Added note on using the default kustomization and what it does (deploys the yaml as is...).

grahamwhaley commented on deployments/gpu_plugin/overlays/nfd_labeled_nodes/add-nodeselector-intel-gpu.yaml:
```diff
@@ -0,0 +1,9 @@
+  template:
+    spec:
+      nodeSelector:
+        feature.node.kubernetes.io/pci-0300_8086.present: "true"
```
nice. I had to go look up NFD and the PCI labelling :-). I presume this detects all our useable Intel GPUs? I wonder if it would also detect GPUs that are too old to run compute on though? I dunno...

askervin replied:
Good point. @uniemimu, do you know if NFD has anything more precise for discovering right nodes?

As long as this does not exclude nodes with gpu_plugin acceleration, it's better than nothing - which would deploy the plugin on all nodes.

mythi replied:
This is nice indeed. Just one concern from QAT perspective: the default label pattern (it's configurable) does not give the PCI IDs. We can only detect "Co-processor" (IIRC)...

mythi replied:
A QAT overlay for dpdk + c62x HW could be:

```
feature.node.kubernetes.io/pci-37c8_8086.present: "true"
feature.node.kubernetes.io/iommu-enabled: "true"
```
and `c6xxvf` for `KERNEL_VF_DRIVERS`.

askervin replied:
@grahamwhaley, GPU plugin itself will detect capabilities of the GPU on the host and expose those as node labels. (This is WIP by @uniemimu.) So `pci-0300_8086` is accurate enough before deploying the plugin.

grahamwhaley replied:
ack, sure, the plugin itself will ultimately decide. np.

askervin commented:
@mythi, this is ok for review. What do you think, should we support NFD labels also in other plugins, as suggested by @uniemimu for GPU?

@rojkov , included kustomization `overlays/default_namespace` here, so kustomizations in PR #271 would not be needed as suggested by @mythi .

mythi commented:
> What do you think, should we support NFD labels also in other plugins, as suggested by @uniemimu for GPU?

Yes we should. Thanks for taking the first step in this PR. Customization makes it simple

mythi commented on deployments/gpu_plugin/overlays/default_namespace/kustomization.yaml:
```diff
@@ -0,0 +1,9 @@
+    version: v1
+    kind: DaemonSet
+    name: intel-gpu-plugin
+  path: daemonset_remove_namespace.json
```
We could just drop it from the original yaml?

rojkov replied:
+1

askervin replied:
Currently all plugins (fpga, gpu, vpu) except for qat_plugin deploy their daemonset to the `kube-system` namespace by default. This is the case also in all device plugins in k8s device plugin examples. So I'd rather prefer moving qat_plugin DaemonSet to `kube-system`, too.

mythi replied:
the problem was that e2e testing using the k8s e2e framework was failing because it wants to set a test specific namespace...

askervin replied:
I see. After this patch PR #271 can use in `test/e2e/gpu/gpu.go:
```
kustomizationYaml = "deployments/gpu_plugin/overlays/no_namespace/kustomization.yaml"
```

mythi replied:
My feeling is that the opposite is the best approach: don't specify the namespace in the default .yamls but the user can decide:

```
$ kubectl create ns qattest
$ kubectl apply -k deployments/qat_plugin/overlay/debug -n qattest
```

whereas this is not possible right now:

```
$ kubectl apply -f deployments/gpu_plugin/gpu_plugin.yaml -n qattest
error: the namespace from the provided object "kube-system" does not match the namespace "qattest". You must pass '--namespace=kube-system' to perform this operation.
```

askervin replied:
Ok.

Do you think we should still provide something like `deployments/gpu_plugin/overlays/namespace_kube-system` so that there's still a way to run the plugin directly in `kube-system` like all the other plugins as well? Or would this be an unnecessary overlay as `kubectl apply -k github.../gpu_plugin -n kube-system` would do the job anyway? I don't know if other usual ways (APIs, bindings) to deploy the plugin offer configuring namespace as easily.

Ok to remove `namespace: kube-system` from the rest of the plugin yamls in later commit(s)?

mythi replied:
> Do you think we should still provide something like deployments/gpu_plugin/overlays/namespace_kube-system
...
> Ok to remove namespace: kube-system from the rest of the plugin yamls in later commit(s)?

+1 from my side

grahamwhaley replied:
I can't think of any reason not to remove the namespaces from the other yamls. As long as we....
- document that they will go in the default namespace (I presume) unless otherwise specified
- ask the other plugin/yaml owners what they think ;-)

askervin commented:
As far as I see, this patch would be ready. Last force-pushed changes: renamed `overlays/default_namespace` to `overlays/no_namespace` and added a missing newline at the end of a file where it was missing.

The discussion resulted in further actions to be taken in next patches:
- `overlays/nfd_labeled_nodes` to be created for `qat_plugin`. It should add two labels to nodeSelector:
   ```
    feature.node.kubernetes.io/pci-37c8_8086.present: "true"
    feature.node.kubernetes.io/iommu-enabled: "true"
   ```
- `qat_plugin` DaemonSet to be deployed in the `kube-system` namespace by default.

askervin commented:
- `namespace: kube-system` removed from the default DaemonSet. (/cc @rojkov )
- `overlays/namespace_kube-system` added
- New commit message:

    gpu_plugin: add kustomizations

    - Default deployment: `kubectl apply -k deployments/gpu_plugin`
    - Default deployment does not specify namespace anymore
      (was: `kube-system`).
    - Variant: deploy only on nodes with Intel GPU label by NFD:
      `kubectl apply -k deployments/gpu_plugin/overlays/nfd_labeled_nodes`
    - Variant: deploy to `kube-system` instead of user-defined namespace
      (or "default"):
      `kubectl apply -k deployments/gpu_plugin/overlays/namespace_kube-system`
    - GPU plugin README updated.

mythi commented on deployments/gpu_plugin/overlays/nfd_labeled_nodes/add-nodeselector-intel-gpu.yaml:
```diff
@@ -0,0 +1,9 @@
+  template:
+    spec:
+      nodeSelector:
+        feature.node.kubernetes.io/pci-0300_8086.present: "true"
```
one problem occured to me: the planned QAT setting would conflict with this. Meaning: we cannot deploy QAT plugin with the same NFD pci setting (`class_vendor` vs `device_vendor`) than GPU... we'll need to re-think this a bit.

mythi commented:
+1 to merge now to unblock #271. We can come back to the NFD labels with QAT.

askervin commented:
@mythi , NFD labels with QAT work started here:

https://github.com/kubernetes-sigs/node-feature-discovery/pull/283
